### PR TITLE
Processing data

### PR DESCRIPTION
**Pull request by Ngufor-emmanuella**
###  In this second half of the mile stone I did the following

Set up associations between models.
Use models to indert data
Create custom methods for post, user, like, and comment models
All linter errors have passed

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,25 +1,274 @@
...
+- FAQ
 
-* Configuration
+After you're finished please remove all the comments and instructions!
+-->
 
-* Database creation
+<div align="center">
+  <!-- You are encouraged to replace this logo with your own! Otherwise you can also remove it. -->
+  <img src="murple_logo.png" alt="logo" width="140"  height="auto" />
```
Comment (FrederickMih) [inline README.md]: - [Optional] @Ngufor-emmanuella this logo image is not working. Let's kindly fix the issue or get rid of the logo. Broken images hurt users' experience.

File: README.md [hunk 2]
```diff
@@ -1,25 +1,274 @@
...
+
+### Usage
+
+To run the project, execute the following command:
+
+  rails server
+
+
+### Run tests
+
```
Comment (FrederickMih) [inline README.md]: - [ ] Hi @Ngufor-emmanuella, you'll be required to provide thorough examples of commands necessary to set up this project locally.

File: README.md [hunk 3]
```diff
@@ -1,25 +1,274 @@
...
+  <ul>
+    <li><a href="https://www.postgresql.org/">PostgreSQL</a></li>
+  </ul>
+</details>
+
+<!-- Features -->
+
+### Key Features <a name="key-features"></a>
+
+> Describe between 1-3 key features of the application.
```
Comment (FrederickMih) [inline README.md]: - [Optional] Always, kindly remove all these descriptions blockquote in your entire README file. The final README.md file should be free from all these directive information from the template.

File: README.md [hunk 4]
```diff
@@ -1,25 +1,274 @@
...
+
+- **[Question_1]**
+
+  - [Answer_1]
+
+- **[Question_2]**
+
+  - [Answer_2]
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
```
Comment (FrederickMih) [inline README.md]: - [Optional] @Ngufor-emmanuella, if you can't complete this section, always get rid of it to render the README file more descriptive.

File: README.md [hunk 5]
```diff
@@ -1,25 +1,274 @@
...
+
+<!-- LICENSE -->
+
+## 📝 License <a name="license"></a>
+
+This project is [MIT](./LICENSE) licensed.
```
Comment (FrederickMih) [inline README.md:213]: - [ ] MIT license link is not working. It's important to always double-check links. Broken links really hurt. This makes work looks unprofessional

File: app/models/post.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+class Post < ApplicationRecord
+  has_many :comments, class_name: 'Comment', foreign_key: :post_id
+  has_many :likes, class_name: 'Like', foreign_key: :post_id
+
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (FrederickMih) [inline app/models/post.rb:5]: - [Optional] @Ngufor-emmanuella, I think you don't need `foreign_key: :author_id`, Rails can always figure out this by convention. **class_name** in the posts table will let Rails know that each of these keys points to a user. Now Rails will look for **author_id** foreign_key in the posts table named after the association.

File: app/models/comment.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :post, class_name: 'Post', foreign_key: :post_id
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (FrederickMih) [inline app/models/comment.rb:3]: - [Optional] If interested check the commend above ⬆️

File: app/models/like.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :post, class_name: 'Post', foreign_key: :post_id
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (FrederickMih) [inline app/models/like.rb:3]: - [Optional] ⬆️

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, class_name: 'Post', foreign_key: :author_id
+  has_many :comments, class_name: 'Comment', foreign_key: :author_id
+  has_many :likes, class_name: 'Like', foreign_key: :author_id
+
+  def recent_posts
+    Post.order(created_at: :desc).includes(:author).limit(3)
```
Comment (FrederickMih) [inline app/models/user.rb:7]: - [ ] A User has many **posts** and you are required to return the 3 most recent **posts** from those **posts** created by the user. Hence I think you should call **order** on those **posts**. I think it'll better works like this:

`posts.order('created_at Desc').limit(3)`

File: app/models/post.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+  def update_post_count
+    author.increment!(:posts_counter)
+  end
+
+  def recent_comments
+    Comment.order(created_at: :desc).limit(5)
```
Comment (FrederickMih) [inline app/models/post.rb:12]: - [ ] We wish to return the five most recent` comments` A post has many `comments` 👍 . Check the comment above for more information: **Hints**

`comments.order('created_at Desc').limit(5)`

File: app/models/comment.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+
+  def update_comment_counter
+    comment.increment!(:comments_counter)
```
Comment (FrederickMih) [inline app/models/comment.rb]: - [ ] We are updating the comments counter for **a post** and not on a `comment`. Comments belong to a **post**

`post.increment!(:comments_counter)`

File: app/models/like.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+
+  def update_likes_counter
+    like.increment!(:likes_counter)
```
Comment (FrederickMih) [inline app/models/like.rb]: - [ ] Let's kindly update the likes counter for a **post** not for a **like**. **Hints**


`post.increment!(:likes_counter)`

**Review by FrederickMih — CHANGES_REQUESTED** [review]
## Hi 🙋 @Ngufor-emmanuella , 


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!




## Highlights </img>


- Correct Gitflow ✔️ 
- PR has a good title and summary ✔️ 
- Linters checks passed ✔️


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

- @Ngufor-emmanuella let's kindly always make sure that our commit messages are clear and descriptive. The commit message [6896cfd](https://github.com/Ngufor-emmanuella/proj-blog-app/pull/2/commits/6896cfde869379ac4a23f50314523a027768e0b8) is not clear enough and contain typos.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md [hunk 13]
```diff
@@ -1,25 +1,274 @@
...
+
+> Describe how a new developer could make use of your project.
+
+To get a local copy up and running, follow these steps.
+
+### Prerequisites
+
+In order to run this project you need:
+
+<!--
```
Comment (TadesseDev) [inline README.md]: - [ ] Dear @Ngufor-emmanuella, As the previous code reviewer mentioned, Your `Getting Started` section is not clear about how a nontechnical user will be able to get started with the project. Please make sure to provide a clear direction for usage. 

Screenshot 2023-02-11 121439

File: README.md (hunk 5 above)
Comment (TadesseDev) [inline README.md]: - [ ] As the previous code reviewer mentioned, I noticed the link for your MIT file is broken, Please make sure to add the MIT license file and update its content in regard to your name and project. Also replace the `[YOUR NAME]` part with your name, and `[APP TYPE]` with your app type.

File: README.md [hunk 14]
```diff
@@ -1,25 +1,274 @@
...
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LICENSE -->
+
+## 📝 License <a name="license"></a>
+
+This project is [MIT](./LICENSE) licensed.
+
+_NOTE: we recommend using the [MIT license](https://choosealicense.com/licenses/mit/) - you can set it up quickly by [using templates available on GitHub](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository). You can also use [any other license](https://choosealicense.com/licenses/) if you wish._
+
```
Comment (TadesseDev) [inline README.md]: - [Optional] Please remove this and the same section from your Readme file, this is just to give you direction on how to make use of `MIT` file and should no more be part of your readme file.

File: app/models/comment.rb [hunk 15]
```diff
@@ -0,0 +1,8 @@
...
+  end
+end
```
Comment (TadesseDev) [inline app/models/comment.rb]: - [ ] Dear @Ngufor-emmanuella, As the previous code reviewer mentioned Comments belong to a post, and the counter should be updated for a specific post. Please make sure to use the `post.increment!(:comments_counter)`, that way the counter for the post will be updated.

File: app/models/like.rb [hunk 16]
```diff
@@ -0,0 +1,8 @@
...
+  end
+end
```
Comment (TadesseDev) [inline app/models/like.rb]: Dear @Ngufor-emmanuella, As the previous code reviewer mentioned Likes belong to a post, and the counter should be updated for a specific post. Please make sure to use the `post.increment!(:likes_counter)`, that way the like counter for the post will be updated.

**Review by TadesseDev — CHANGES_REQUESTED** [review]
### Hi @Ngufor-emmanuella 

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

199733072-bdacca19-8df9-43a7-be93-3aff2f1c15c8

 ## Highlights
 - [ ] Good job making most of the changes requested by the previous reviewers. 🙏


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by emmanuellekamwa — APPROVED** [review]
**STATUS: APPROVED ✔🍾🎉**
Hello @Ngufor-emmanuella 👋
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- You documented your work professionally.✔
- You asked for the changes in a polite and friendly way.
- Your linters are passing all the checks.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Ngufor-emmanuella** [comment]
thank you

**Pull request merged**